reports/tests.py

```diff
@@ -0,0 +1,521 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""Tests module for the invoices reports"""
+
+# Imports #####################################################################
+from collections import defaultdict
+from datetime import timedelta, datetime
+from unittest import mock
+
+from django.conf import settings
+from django.core.urlresolvers import reverse
+from django.test import TestCase, override_settings
+from django.utils import timezone
+
+import ddt
+
+from instance.models.appserver import Status as AppServerStatus
+from instance.models.openedx_instance import OpenEdXInstance
+from instance.tests.base import WithUserTestCase
+from instance.tests.models.factories.openedx_appserver import make_test_appserver
+from instance.tests.models.factories.openedx_instance import OpenEdXInstanceFactory
+from pr_watch.models import WatchedPullRequest
+from pr_watch.tests.factories import WatchedForkFactory
+from reports.helpers import (
+    _normalize_date,
+    generate_charge_details,
+    generate_charges,
+    generate_watched_forks_instances,
+    get_billing_dates,
+    get_instance_charges,
+)
+from userprofile.factories import OrganizationFactory
+
+# Tests #####################################################################
+
+
+class ReportViewTestCase(WithUserTestCase):
+    """
+    Tests for reports views functionality
+    """
+    url = 'reports:report'
+
+    def setUp(self):
+        super(ReportViewTestCase, self).setUp()
+
+        self.username = 'user3'
+        self.password = 'pass'
+        self.organization = OrganizationFactory.create(
+            name='Organization Name',
+            github_handle='organization_github_handle'
+        )
+
+    def test_report_view_not_superuser(self):
+        """
+        Users who are not logged in or logged in but don't have
+        superuser credentials should not be able to see the page.
+        """
+        invoice_url = reverse(self.url, kwargs={
+            'organization': 'dummy_org',
+            'year': '2018',
+            'month': '9'
+        })
+        expected_redirect = '{}?next={}'.format(reverse('registration:login'), invoice_url)
+
+        # Anonymous user
+        response = self.client.get(invoice_url)
+        self.assertRedirects(
+            response, expected_redirect,
+            status_code=302,
+            target_status_code=200,
+            msg_prefix=''
+        )
+
+        # Normal User
```

Comment:
Could you use ddt to avoid repeating code? There are [other tests that try several unauthorized users](https://github.com/open-craft/opencraft/blob/fbd28cea4962056a70078098f90a0f5e7f8b01f5/instance/tests/api/test_openedx_appserver.py#L111). Could you also add user4 (instance manager) to the test? (instance managers shouldn't see the reports; it's important to include this in the checks).